File: api/src/main/java/dev/vibris/api/CancellationToken.java
package dev.vibris.api;

import java.util.concurrent.CancellationException;
import java.util.concurrent.atomic.AtomicBoolean;

@FunctionalInterface
public interface CancellationToken {
    boolean isCancellationRequested();

    default void throwIfCancellationRequested() {
        if (isCancellationRequested()) throw new CancellationException("Vibris operation was cancelled");
    }

    static CancellationToken none() {
        return () -> false;
    }

    static Source source() {
        return new Source();
    }

    final class Source {
        private final AtomicBoolean cancelled = new AtomicBoolean();
        private final CancellationToken token = cancelled::get;

        private Source() {
        }

        public CancellationToken token() {
            return token;
        }

        public boolean cancel() {
            return cancelled.compareAndSet(false, true);
        }
    }
}

File: api/src/main/kotlin/dev/vibris/api/CancellationToken.kt
package dev.vibris.api

import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicBoolean

fun interface CancellationToken {
    fun isCancellationRequested(): Boolean

    fun throwIfCancellationRequested() {
        if (isCancellationRequested()) {
            throw CancellationException("Vibris operation was cancelled")
        }
    }

    class Source private constructor() {
        private val cancelled = AtomicBoolean()
        private val token = CancellationToken { cancelled.get() }

        fun token(): CancellationToken = token

        fun cancel(): Boolean = cancelled.compareAndSet(false, true)

        companion object {
            internal fun create(): Source = Source()
        }
    }

    companion object {
        @JvmStatic
        fun none(): CancellationToken = CancellationToken { false }

        @JvmStatic
        fun source(): Source = Source.create()
    }
}